docs/architecture/design-proposals/third-party-add-repository.md

```diff
@@ -0,0 +1,96 @@
+# Add a repository and trigger a chart deployment from an external source
+
+## Objective
+
+Enable an external user-interface for chart repositories, such as Harbor, to trigger the addition of a repository in Kubeapps or trigger the deployment of a chart directly into the Kubeapps UI.
+
+## Motivation
+
+Kubeapps focuses presenting a catalog of apps from multiple repositories and enabling easy deployment of those apps. Kubeapps does not (and should not) focus on the management of app repositories nor populating them with apps.
+
+Other projects, such as Harbor, focus specifically on the management of repositories for cloud-native resources such as images and apps. These project don't necessarily focus on the user experience for deploying these apps or images.
+
+Enabling these projects to integrate with Kubeapps enables users to move seamlessly from uploading a new app to a repository through to testing the deployment of that same app in their configured Kubeapps installation. Or from preparing a repository with specific apps to enabling users in Kubeapps to access those apps.
+
+## Goals and non goals
+
+* Enable 3rd party sites to provide an "Add repository to Kubeapps" action which results in the Kubeapps installation scanning and including the repository in its catalog.
+* As a consequence, the 3rd party can then link to directly deploy a chart from a repository in the Kubeapps UI.
+
+## Assumptions
+ - The same OpenID Connect identity provider must be used for both the 3rd party (such as Harbor) and Kubeapps (and by implication, the Kubernetes cluster).
```

Comment:
AFAIK, the default scenario is to have Kubeapps/Harbor configured with username and password or the token, should we support this scenario?

Reply:
AFAIUI, Harbor only has username/password (called "database"), LDAP or OpenID Connect (see [auth section of the user guide](https://github.com/goharbor/harbor/blob/master/docs/user_guide.md#managing-authentication)).

I'm keen to limit this initial integration to OpenID Connect for a couple of reasons:
 * I'm actually keen for us (Kubeapps) to move towards recommending OpenID Connect as the default rather than token, with the k8s service account token being an option for a quick setup and testing, but not recommended for production use.
 * It makes the integration trivial - since an `id_token` from harbor is valid for Kubeapps/K8s if it's issued by the same identity provider).
 * It is a common auth across all three (harbor, kubeapps and the k8s api server).

Note that it doesn't exclude the systems using separate auth systems... a cookie credential could be used from harbor to Kubeapps as mentioned below (ie. a [JS Fetch request with the `include` option for cross-origin cookies](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API/Using_Fetch#Sending_a_request_with_credentials_included) will send the kubeapps cookie once a user is auth'd), but this does not work for a token in localStorage - available only to the document's origin. So, unless we were wanting users to set a K8s service account token in the 3rd party app (?!), I think we're dependent on Kubeapps auth proxy to be able to have the 3rd party do api requests?

Reply:
I see. I agree with the reasoning. Maybe it's not something to discuss in this document but I wanted to clarify what happen if OIDC is not configured. I am not saying that we need to support it but at least plan to have that case covered (documentation or a proper error may be enough).